lib/engine/auto_router.rb

```diff
@@ -28,21 +30,44 @@ def compute(corporation, **opts)
         ]
       end
 
+      # Add a per-node timeout to ensure we don't spend ALL time on first node's paths in huge maps,
+      # which can cause all train routes to conflict with each other
+      node_timeout = path_timeout / nodes.size
+
       now = Time.now
 
       skip_paths = static.flat_map(&:paths).to_h { |path| [path, true] }
 
+      train_routes = Hash.new { |h, k| h[k] = [] }    # map of train to route list
+      path_abort = Hash.new { |h, k| h[k] = false }   # each train has opportunity to abort a branch of the path walk tree
+      hexside_bits = Hash.new { |h, k| h[k] = 0 }     # map of hexside_id to bit number
+      @nextbit = 0
+
       nodes.each do |node|
         if Time.now - now > path_timeout
           puts 'Path timeout reached'
           break
         else
-          puts "Path search: #{nodes.index(node)} / #{nodes.size}"
+          puts "Path search: #{nodes.index(node)} / #{nodes.size} - paths starting from #{node.hex.name}"
         end
 
+        node_now = Time.now
+
+        counter = 0
+        node_abort = false
+
         node.walk(corporation: corporation, skip_paths: skip_paths) do |_, vp|
+          next :abort if node_abort
+
           paths = vp.keys
 
+          if Time.now - node_now > node_timeout
```

Comment:
would this be simplified as node_abort = Time.now - node_now > node_timeout
no need to do next :abort since it's already done in the first part of the loop